graphbolt/src/random_engine.cc

```diff
@@ -0,0 +1,125 @@
+          // otherwise get a new value until we haven't unique range of
+          // elements.
+          auto it = std::find(out, b, *b);
+          if (it != b) continue;
+          ++b;
+        }
+      } else {
+        // use hash set
+        // In the best scenario, time complexity is O(num), i.e., no conflict.
+        //
+        // Let k be num / population, the expected number of extra sampling
+        // steps is roughly k^2 / (1-k) * population, which means in the worst
+        // case scenario, the time complexity is O(population^2). In practice,
+        // we use 1/10 since std::unordered_set is pretty slow.
+        std::unordered_set<IdxType> selected;
+        while (static_cast<IdxType>(selected.size()) < num) {
+          selected.insert(RandInt(population));
+        }
+        std::copy(selected.begin(), selected.end(), out);
+      }
+
+    } else {
+      // In this case, `num >= population / 10`. To reduce the computation
+      // overhead, we should reduce the number of random number generations.
+      // Even though reservior algorithm is more memory effficient (it has
+      // O(num) memory complexity), it generates O(population) random numbers,
+      // which is computationally expensive. This algorithm has memory
+      // complexity of O(population) but generates much fewer random numbers
+      // O(num). In the case of `num >= population/10`, we don't need to worry
+      // about memory complexity because `num` is usually small. So is
+      // `population`. Allocating a small piece of memory is very efficient.
+      std::vector<IdxType> seq(population);
+      for (size_t i = 0; i < seq.size(); i++) seq[i] = i;
+      for (IdxType i = 0; i < num; i++) {
+        IdxType j = RandInt(i, population);
+        std::swap(seq[i], seq[j]);
+      }
+      // Save the randomly sampled numbers.
+      for (IdxType i = 0; i < num; i++) {
+        out[i] = seq[i];
+      }
+    }
+  }
+}
+
+template void RandomEngine::UniformChoice<int32_t>(
+    int32_t num, int32_t population, int32_t* out, bool replace);
+template void RandomEngine::UniformChoice<int64_t>(
+    int64_t num, int64_t population, int64_t* out, bool replace);
+
+// Perform a biased choice using probabilities for each element with optional
+// replacement.
+template <typename IdxType, typename ProbType>
+void RandomEngine::Choice(
+    IdxType num, ProbType* prob, IdxType len, IdxType* out, bool replace) {
+  std::discrete_distribution<IdxType> dist(prob, prob + len);
+  if (replace) {
+    for (auto i = 0; i < num; i++) out[i] = dist(rng_);
+  } else {
+    // TODO: Naive implementation, will be changed in the future.
+    std::vector<ProbType> prob_vec(prob, prob + len);
+    for (auto i = 0; i < num; i++) {
+      out[i] = dist(rng_);
+      prob_vec[out[i]] = 0;
+      dist =
+          std::discrete_distribution<IdxType>(prob_vec.begin(), prob_vec.end());
+    }
+  }
+}
+
+template void RandomEngine::Choice<int32_t, float>(
```

Comment:
Personally I don't like to list instantiated templates here which looks tedious. why not let compiler instantiate during compiling?

Reply:
you mean  `RandomEngine::Choice()` is called in python level? not in c++ file? so it will not be instantiated during compiling graphbolt.so?